File: editors/vscode/src/lsp.ts
// The language server client.
//
// Hand-rolled rather than pulled in through `vscode-languageclient`: the
// server speaks a small, fixed subset, and a full client library would be a
// large dependency to carry for it.
//
// Everything the server offers is asked for here. It used to be only the
// notifications -- open, change, close -- with the reply-carrying half of the
// protocol unimplemented, so completion, hover and go-to-definition were
// advertised by the server, wired to nothing, and simply did not happen.

import { type ChildProcess, spawn } from "node:child_process";
import * as vscode from "vscode";
import { MessageReader, completionKind, completionPrefixStart, frame, markdownOf, replyId } from "./pure";

/** A whole-document replacement, which is the only edit the server sends. */
interface TextEdit {
	range: {
		start: { line: number; character: number };
		end: { line: number; character: number };
	};
	newText: string;
}

interface CompletionItem {
	label: string;
	kind?: number;
	detail?: string;
	documentation?: unknown;
}

interface Diagnostic {
	range: {
		start: { line: number; character: number };
		end: { line: number; character: number };
	};
	severity?: number;
	message: string;
}

/** LSP severities are 1-4; VS Code's enum runs the other way. */
export function severityOf(n: number | undefined): vscode.DiagnosticSeverity {
	switch (n) {
		case 2:
			return vscode.DiagnosticSeverity.Warning;
		case 3:
			return vscode.DiagnosticSeverity.Information;
		case 4:
			return vscode.DiagnosticSeverity.Hint;
		default:
			return vscode.DiagnosticSeverity.Error;
	}
}

/**
 * The standalone server, as it was installed beside `run` before the runner
 * grew a `:lsp` of its own.
 *
 * Only a fallback, and only for the window where the two can disagree: the
 * marketplace updates this extension on its own, while `run` is updated by
 * hand, so a client asking for `run :lsp` can meet a runner that has never
 * heard of it. That runner came with a `runfile-lsp` next to it, which is
 * still on PATH and still answers. Removable once no supported `run` predates
 * the subcommand.
 */
const LEGACY_SERVER = "runfile-lsp";

export class LanguageClient implements vscode.Disposable {
	private child?: ChildProcess;
	private reader = new MessageReader();
	private readonly diagnostics = vscode.languages.createDiagnosticCollection("runfile");
	private readonly subs: vscode.Disposable[] = [];
	private nextId = 1;
	/** Requests waiting on a reply, by id. */
	private readonly pending = new Map<number, (result: unknown) => void>();
	/** Whether the server has said anything at all -- see [`LEGACY_SERVER`]. */
	private answered = false;
	private fellBack = false;
	private disposed = false;

	constructor(
		private readonly command: string,
		private readonly args: string[],
		private readonly log: vscode.OutputChannel,
	) {}

	start(): void {
		this.spawnServer(this.command, this.args);
		// Registered once, outside `spawnServer`: a fallback replaces the
		// process, not this client, and subscribing a second time would send
		// every notification twice.
		this.subs.push(
			vscode.workspace.onDidOpenTextDocument((d) => this.didOpen(d)),
			vscode.workspace.onDidChangeTextDocument((e) => this.didChange(e.document)),
			vscode.workspace.onDidCloseTextDocument((d) => this.didClose(d)),
		);
	}

	private spawnServer(command: string, args: string[]): void {
		// A fresh reader: a half-written frame from a process that died is not
		// the start of the next one's first message.
		this.reader = new MessageReader();
		try {
			this.child = spawn(command, args, { stdio: ["pipe", "pipe", "pipe"] });
		} catch (e) {
			this.log.appendLine(`language server did not start: ${String(e)}`);
			this.child = undefined;
			return;
		}
		this.child.on("error", (e) => {
			// Not installed is the common case, and it must not be a popup.
			this.log.appendLine(`language server unavailable (${command}): ${e.message}`);
			this.child = undefined;
			this.fallBack("could not be started");
		});
		// A runner with no `:lsp` prints its refusal here and exits 1.
		this.child.on("exit", (code) => this.fallBack(`exited with code ${code}`));
		this.child.stderr?.on("data", (b: Buffer) => this.log.append(b.toString()));
		this.child.stdout?.on("data", (b: Buffer) => {
			for (const message of this.reader.push(b)) {
				this.answered = true;
				this.receive(message);
			}
		});

		this.send({ jsonrpc: "2.0", id: this.nextId++, method: "initialize", params: {} });
		this.send({ jsonrpc: "2.0", method: "initialized", params: {} });

		// Whatever is already open when the extension activates never fires
		// `onDidOpen`, so it is opened explicitly. Also what re-syncs the
		// documents onto a fallback server, which missed them the first time.
		for (const doc of vscode.workspace.textDocuments) {
			this.didOpen(doc);
		}
	}

	/**
	 * Try [`LEGACY_SERVER`] once, if the configured one never said anything.
	 *
	 * A server that answered and then died is a crash, not a version
	 * mismatch, and reaching for a different binary would hide it.
	 */
	private fallBack(why: string): void {
		if (this.answered || this.fellBack || this.disposed) {
			return;
		}
		this.fellBack = true;
		this.log.appendLine(`${[this.command, ...this.args].join(" ")} ${why}; trying ${LEGACY_SERVER}`);
		this.spawnServer(LEGACY_SERVER, []);
	}

	dispose(): void {
		this.disposed = true;
		for (const s of this.subs) {
			s.dispose();
		}
		this.diagnostics.dispose();
		this.child?.kill();
	}

	private send(message: unknown): void {
		this.child?.stdin?.write(frame(message));
	}

	/**
	 * Send a request and wait for its reply.
	 *
	 * Times out rather than waiting forever: this runs on save, and a server
	 * that has wedged must not take the editor's save with it.
	 */
	private request(method: string, params: unknown, timeoutMs = 2000): Promise<unknown> {
		if (!this.child) {
			return Promise.resolve(null);
		}
		const id = this.nextId++;
		return new Promise((resolve) => {
			const done = (result: unknown) => {
				clearTimeout(timer);
				this.pending.delete(id);
				resolve(result);
			};
			const timer = setTimeout(() => done(null), timeoutMs);
			this.pending.set(id, done);
			this.send({ jsonrpc: "2.0", id, method, params });
		});
	}

	/**
	 * What may be written at this position: properties after a `.`, functions
	 * and sources, `RUN.` keys, target names after `run `.
	 *
	 * The range is given explicitly. VS Code's own idea of a word ends at a
	 * `-` and a `:`, which would make `.env-file` complete to `.env-env-file`.
	 */
	async completion(doc: vscode.TextDocument, pos: vscode.Position): Promise<vscode.CompletionItem[]> {
		const result = (await this.position("textDocument/completion", doc, pos)) as {
			items?: CompletionItem[];
		} | null;
		const items = result?.items;
		if (!Array.isArray(items)) {
			return [];
		}
		const line = doc.lineAt(pos.line).text;
		const range = new vscode.Range(
			pos.line,
			completionPrefixStart(line, pos.character),
			pos.line,
			pos.character,
		);
		return items.map((i) => {
			const item = new vscode.CompletionItem(i.label, completionKind(i.kind) as vscode.CompletionItemKind);
			item.detail = i.detail;
			const doc = markdownOf(i.documentation);
			if (doc) {
				item.documentation = new vscode.MarkdownString(doc);
			}
			item.range = range;
			return item;
		});
	}

	async hover(doc: vscode.TextDocument, pos: vscode.Position): Promise<vscode.Hover | undefined> {
		const result = (await this.position("textDocument/hover", doc, pos)) as { contents?: unknown } | null;
		const text = markdownOf(result?.contents);
		return text ? new vscode.Hover(new vscode.MarkdownString(text)) : undefined;
	}

	async definition(doc: vscode.TextDocument, pos: vscode.Position): Promise<vscode.Location | undefined> {
		const result = (await this.position("textDocument/definition", doc, pos)) as {
			uri?: string;
			range?: TextEdit["range"];
		} | null;
		if (!result?.uri || !result.range) {
			return undefined;
		}
		const r = result.range;
		return new vscode.Location(
			vscode.Uri.parse(result.uri),
			new vscode.Range(r.start.line, r.start.character, r.end.line, r.end.character),
		);
	}

	/** A request about a place in a document, which is most of them. */
	private position(method: string, doc: vscode.TextDocument, pos: vscode.Position): Promise<unknown> {
		if (!this.isRunfile(doc)) {
			return Promise.resolve(null);
		}
		return this.request(method, {
			textDocument: { uri: doc.uri.toString() },
			position: { line: pos.line, character: pos.character },
		});
	}

	/**
	 * The edits that put a document into the one shape there is -- the same
	 * `run :format` produces, since both call the same formatter.
	 *
	 * An empty list when nothing needs changing, so saving a clean file marks
	 * nothing dirty; nothing at all when the document does not parse, because
	 * a file is unfinished for most of the time it is being written.
	 */
	async format(doc: vscode.TextDocument): Promise<vscode.TextEdit[]> {
		if (!this.isRunfile(doc)) {
			return [];
		}
		const result = (await this.request("textDocument/formatting", {
			textDocument: { uri: doc.uri.toString() },
			options: { tabSize: 4, insertSpaces: false },
		})) as TextEdit[] | null;
		if (!Array.isArray(result)) {
			return [];
		}
		// The server sends one edit past the last line; clamp it to what the
		// document actually has, which is what VS Code will accept.
		const full = new vscode.Range(0, 0, doc.lineCount, 0);
		return result.map((e) => vscode.TextEdit.replace(full, e.newText));
	}

	private receive(message: unknown): void {
		const m = message as {
			id?: number;
			result?: unknown;
			method?: string;
			params?: { uri?: string; diagnostics?: Diagnostic[] };
		};
		const id = replyId(message);
		if (id !== undefined) {
			this.pending.get(id)?.(m.result ?? null);
			return;
		}
		if (m.method !== "textDocument/publishDiagnostics" || !m.params?.uri) {
			return;
		}
		const uri = vscode.Uri.parse(m.params.uri);
		this.diagnostics.set(
			uri,
			(m.params.diagnostics ?? []).map((d) => {
				const range = new vscode.Range(
					d.range.start.line,
					d.range.start.character,
					d.range.end.line,
					d.range.end.character,
				);
				const out = new vscode.Diagnostic(range, d.message, severityOf(d.severity));
				out.source = "runfile";
				return out;
			}),
		);
	}

	private isRunfile(doc: vscode.TextDocument): boolean {
		return doc.uri.scheme === "file" && doc.uri.fsPath.endsWith(".run");
	}

	private didOpen(doc: vscode.TextDocument): void {
		if (!this.isRunfile(doc)) {
			return;
		}
		this.send({
			jsonrpc: "2.0",
			method: "textDocument/didOpen",
			params: {
				textDocument: {
					uri: doc.uri.toString(),
					languageId: "runfile",
					version: doc.version,
					text: doc.getText(),
				},
			},
		});
	}

	private didChange(doc: vscode.TextDocument): void {
		if (!this.isRunfile(doc)) {
			return;
		}
		// Full sync, matching what the server advertises.
		this.send({
			jsonrpc: "2.0",
			method: "textDocument/didChange",
			params: {
				textDocument: { uri: doc.uri.toString(), version: doc.version },
				contentChanges: [{ text: doc.getText() }],
			},
		});
	}

	private didClose(doc: vscode.TextDocument): void {
		if (!this.isRunfile(doc)) {
			return;
		}
		this.send({
			jsonrpc: "2.0",
			method: "textDocument/didClose",
			params: { textDocument: { uri: doc.uri.toString() } },
		});
	}
}

File: crates/runfile-cli/tests/lsp.rs
//! `run :lsp` -- the language server as the shipped binary serves it.
//!
//! `runfile-lsp/tests/protocol.rs` exercises the `Server` type; nothing there
//! would notice a `:lsp` arm that never dispatched, a runner that printed a
//! line of its own onto the stream LSP framing owns, or stdout that never
//! flushed. This lives beside the CLI's other tests because the subcommand is
//! now what ships -- there is no second binary to start.

use std::io::Write;
use std::process::{Command, Stdio};

use serde_json::{Value, json};

use runfile_lsp::rpc::{read_message, write_message};

/// Feed a conversation to the real binary and collect its replies.
///
/// Everything is written and stdin closed before reading, so neither side can
/// block on the other: the server reads until end of file.
fn talk(messages: &[Value]) -> Vec<Value> {
	let mut input = Vec::new();
	for m in messages {
		write_message(&mut input, m).expect("frame");
	}
	let mut child = Command::new(env!("CARGO_BIN_EXE_run"))
		.arg(":lsp")
		.stdin(Stdio::piped())
		.stdout(Stdio::piped())
		.stderr(Stdio::null())
		.spawn()
		.expect("spawn run :lsp");
	child.stdin.take().expect("stdin").write_all(&input).expect("write");
	let out = child.wait_with_output().expect("wait");
	assert!(out.status.success(), "server exited with {}", out.status);

	let mut r = std::io::BufReader::new(&out.stdout[..]);
	let mut replies = Vec::new();
	while let Ok(v) = read_message(&mut r) {
		replies.push(v);
	}
	replies
}

fn did_open(uri: &str, text: &str) -> Value {
	json!({
		"jsonrpc": "2.0",
		"method": "textDocument/didOpen",
		"params": {"textDocument": {"uri": uri, "languageId": "runfile", "text": text}},
	})
}

#[test]
fn run_lsp_answers_initialize_and_reports_a_problem() {
	let replies = talk(&[
		json!({"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": {}}),
		did_open("file:///x/runfiles/a.run", ".wach = \"y\"\n$ true\n"),
	]);
	assert_eq!(replies.len(), 2, "{replies:?}");
	assert_eq!(replies[0]["id"], 1);
	assert_eq!(replies[0]["result"]["serverInfo"]["name"], "runfile-lsp");

	let diagnostics = replies[1]["params"]["diagnostics"].as_array().expect("diagnostics");
	assert_eq!(diagnostics.len(), 1, "{diagnostics:?}");
	assert!(
		diagnostics[0]["message"].as_str().unwrap().contains(".watch"),
		"{diagnostics:?}"
	);
}

#[test]
fn run_lsp_exits_zero_when_the_client_says_goodbye() {
	// An editor closing down must not leave a failing process behind.
	let replies = talk(&[
		json!({"jsonrpc": "2.0", "id": 1, "method": "shutdown"}),
		json!({"jsonrpc": "2.0", "method": "exit"}),
	]);
	assert_eq!(replies.len(), 1);
	assert!(replies[0]["result"].is_null());
}

#[test]
fn run_lsp_puts_nothing_of_its_own_on_stdout() {
	// The risk the subcommand introduces that a separate binary did not have:
	// `run` is a program that talks. An announcement, a `[runfile]` line, a
	// usage banner -- anything ahead of the first header desynchronises the
	// client for the rest of the session, and it would do so in a directory
	// with a project in it, which is every directory an editor opens.
	let dir = tempfile::tempdir().expect("tempdir");
	std::fs::create_dir(dir.path().join("runfiles")).expect("runfiles");
	std::fs::write(dir.path().join("runfiles/build.run"), "$ true\n").expect("target");

	let mut input = Vec::new();
	write_message(&mut input, &json!({"jsonrpc": "2.0", "id": 1, "method": "shutdown"})).expect("frame");
	write_message(&mut input, &json!({"jsonrpc": "2.0", "method": "exit"})).expect("frame");

	let mut child = Command::new(env!("CARGO_BIN_EXE_run"))
		.arg(":lsp")
		.current_dir(dir.path())
		.stdin(Stdio::piped())
		.stdout(Stdio::piped())
		.stderr(Stdio::null())
		.spawn()
		.expect("spawn run :lsp");
	child.stdin.take().expect("stdin").write_all(&input).expect("write");
	let out = child.wait_with_output().expect("wait");

	assert!(
		out.stdout.starts_with(b"Content-Length:"),
		"stdout opens with something other than a frame: {:?}",
		String::from_utf8_lossy(&out.stdout[..out.stdout.len().min(120)])
	);
}
